Python 3 support for ShinySDR

quentinmit opened the pull request:
I can't promise everything works but all the basic functionality seems to be there and working in my tests. I kept each logical change as a separate commit so you can easily review them. I believe all of these changes are backwards-compatible to Python 2.7. Several of the tests are failing but it seems like that was true before these commits as well. I can start trying to knock off the tests in this PR if you'd like, or I can leave that for another PR.

quentinmit committed:
Update setup.py for Python 3

quentinmit committed:
Update testutil.py for Python 3

quentinmit committed:
More Python 3 support

quentinmit committed:
Work around Twisted Python 3.8 compatibility issue

quentinmit committed:
Call putChild with byte strings for Python 3 compatibility

quentinmit committed:
Fix make_websocket_url to use strings on Python 3

quentinmit committed:
Filenames and MIME types are strings in Python 3

quentinmit committed:
Return JSON as utf-8

quentinmit committed:
Parse WebSocket path correctly on Python 3

quentinmit committed:
Python blocks must have a reference held on the Python side as long as they are part of a flowgraph

quentinmit committed:
Monkey-patch txWS to fix it on Python 3

quentinmit committed:
Fix BlockResource for Python 3

quentinmit committed:
sort can fail on Python 3 if objects are not comparable

quentinmit committed:
Iterators no longer have a next() method in Python 3

quentinmit committed:
Filenames are not bytes in Py3

kpreid commented:
My computer(s) are a bit in disarray right now, so I may be a while giving this a proper code review. So, thanks for doing this, and I'll give you some quick cheap observations. Unfortunately, the Travis CI integration seems to have not posted results here (I may have reauthorized it now, but I'm not sure) but [here they are](https://travis-ci.org/github/kpreid/shinysdr/builds/710322886). There are two simple lints:

```
shinysdr/i/network/webapp.py:79:1: E305 expected 2 blank lines after class or function definition, found 0
shinysdr/i/network/export_http.py:35:1: F401 'shinysdr.i.pycompat.defaultstr' imported but unused
```

And there are several test failures in the Python 2.7 environment, all looking like they have to do with details of web serving (probably str/bytes/unicode confusion). It's okay to go from failing to failing on Python 3, but the tests must remain passing under Python 2 (until we drop Python 2 support, which will come no sooner than adding GNU Radio 3.8 support).

quentinmit commented:
> My computer(s) are a bit in disarray right now, so I may be a while giving this a proper code review. So, thanks for doing this, and I'll give you some quick cheap observations. Unfortunately, the Travis CI integration seems to have not posted results here (I may have reauthorized it now, but I'm not sure) but [here they are](https://travis-ci.org/github/kpreid/shinysdr/builds/710322886). There are two simple lints:
> 
> ```
> shinysdr/i/network/webapp.py:79:1: E305 expected 2 blank lines after class or function definition, found 0
> shinysdr/i/network/export_http.py:35:1: F401 'shinysdr.i.pycompat.defaultstr' imported but unused
> ```
> 
> And there are several test failures in the Python 2.7 environment, all looking like they have to do with details of web serving (probably str/bytes/unicode confusion). It's okay to go from failing to failing on Python 3, but the tests must remain passing under Python 2 (until we drop Python 2 support, which will come no sooner than adding GNU Radio 3.8 support).

I don't have a good Python 2 GR environment anymore. I guess I could see how many tests run without being able to import gnuradio. I'll take a stab at trying to fix the bytes confusion.

I didn't mention it in the PR description but this also runs on GNU Radio 3.8.

kpreid commented:
OK, when I have my GR development environment back, I'll see if I can contribute any tweaks to get it working on both &mdash; or perhaps merge whatever subset of these commits are straightforwardly compatible.

quentinmit committed:
Handle application/json headers with and without charset specified

quentinmit committed:
Address lint errors

quentinmit commented:
> OK, when I have my GR development environment back, I'll see if I can contribute any tweaks to get it working on both — or perhaps merge whatever subset of these commits are straightforwardly compatible.

I just pushed a fix to the only test that I could find broken on 2.7 (`shinysdr.i.test_db.TestDatabaseResource`). And the lint fixes. With that I get:

```
Ran 547 tests in 7.796s

PASSED (skips=12, successes=535)
```

There are definitely still test failures left on Python 3.8, though.

I don't know how disarrayed your computers are right now, but if you have a computer that can run Docker you can pull the `w1xm/shinysdr` Docker image if you want a Python 3.8 + GR 3.8 + ShinySDR dev environment. There's also `w1xm/base:python2` if you want GR 3.8 + Python 2.7, though it doesn't have ShinySDR preinstalled.

quentinmit committed:
More bytes/str changes that fix tests on Python 3.8

quentinmit commented:
I can't leave it alone, so I ran trial3 inside docker and got it down to:

```
Ran 547 tests in 11.684s

FAILED (skips=5, failures=3, successes=539)
```

trial2 is still passing as well, and running it on Py3.8 + GR3.8 seems to cursorily work. As part of the most recent commit I think I fixed the OsmoSDR source (we don't use it, though, so I can't be sure).

The only remaining failing tests on Python 3.8 are:

```
shinysdr.i.network.test_export_ws.TestStateStream.test_two_references
shinysdr.i.test_main.TestMain.test_minimal
shinysdr.i.test_main.TestMain.test_persistence
```

AFAICT the last two (and maybe also the first?) appear to be testing the order of items in a dictionary, which was not defined behavior and changed in Python 3.7.

quentinmit committed:
Fix lint

quentinmit commented:
@kpreid Hey, this has been sitting for a couple weeks... any chance to look at this yet?

kpreid commented on setup.py:
```diff
@@ -20,7 +20,10 @@
 import os.path
 import subprocess
-import urllib
+try:
+    from urllib.request import urlretrieve
+except ImportError:
+    from urllib import urlretrieve
```
Add a comment indicating this is for version compatibility so we can strip it out later.

quentinmit replied:
Done

kpreid commented on shinysdr/i/blocks.py:
```diff
@@ -302,8 +302,8 @@ def __do_connect(self):
         
-        fft_sink = self.__fft_cell.create_sink_internal(numpy.dtype((numpy.int8, output_length)))
-        scope_sink = self.__scope_cell.create_sink_internal(numpy.dtype(('c8', self.__time_length)))
+        self.__fft_sink = self.__fft_cell.create_sink_internal(numpy.dtype((numpy.int8, output_length)))
+        self.__scope_sink = self.__scope_cell.create_sink_internal(numpy.dtype(('c8', self.__time_length)))
```
Why are these being made attributes?

quentinmit replied:
Letting Python garbage collect them while they're still in the flowgraph causes crashes now.

kpreid commented on shinysdr/i/db.py:
```diff
@@ -186,11 +186,11 @@ def __init__(self, db, instantiate):
     
     def render_GET(self, request):
-        request.setHeader(b'Content-Type', b'application/json')
+        request.setHeader(b'Content-Type', b'application/json; charset=utf-8')
```
application/json does not have a charset parameter; this change and its other instances below should be reverted. https://tools.ietf.org/html/rfc7159

quentinmit replied:
I get what the RFC says, but I believe this was in fact necessary to make it work because something else started setting/requiring it. I'll try to reproduce the error so I can post it here.

(I certainly didn't spontaneously decide to add this all over the place!)

quentinmit replied:
I can't reproduce whatever error I saw (and this was months ago so I've long since forgotten). I reverted these changes for now as a separate commit so I can easily re-add them if/when I figure out why they were necessary in the first place.

kpreid commented on shinysdr/i/network/export_http.py:
```diff
@@ -71,19 +70,20 @@ def __init__(self, block, wcommon, deleteSelf):
         self.__element = _BlockHtmlElement(wcommon)
     
     def getChild(self, path, request):
+        name = path.decode()
```
This will use the default encoding when on Python 2, but should always be UTF-8.

quentinmit replied:
Done

kpreid commented on shinysdr/i/network/export_http.py:
```diff
@@ -71,19 +70,20 @@ def __init__(self, block, wcommon, deleteSelf):
                     self._blockCells[key] = cell
                 else:
-                    self.putChild(key, ValueCellResource(cell, self.__wcommon))
+                    self.putChild(key.encode(), ValueCellResource(cell, self.__wcommon))
```
This will use the default encoding when on Python 2, but should always be UTF-8.

quentinmit replied:
Done

kpreid commented on shinysdr/i/network/session_http.py:
```diff
@@ -38,23 +38,23 @@ def __init__(self, session, wcommon, read_only_dbs, writable_db):
         
         # Exported radio control objects
-        self.putChild(CAP_OBJECT_PATH_ELEMENT, BlockResource(session, wcommon, _not_deletable))
+        self.putChild(CAP_OBJECT_PATH_ELEMENT.encode(), BlockResource(session, wcommon, _not_deletable))
```
How about using b'' for the constant value since it's always used in this context? If that doesn't work, I'd rather see .encode('utf-8') here because it's the right thing for all URL components, rather than depending on the default encoding under Python 2.

quentinmit replied:
I'm not sure I actually buy that utf-8 is correct; arguably it should be utf-8 + urlencode, but since it's a constant that's pure ASCII it's kind of moot. Done.

kpreid commented on shinysdr/i/network/session_http.py:
```diff
@@ -38,23 +38,23 @@ def __init__(self, session, wcommon, read_only_dbs, writable_db):
         
         # Standard audio-file-over-HTTP audio stream (the ShinySDR web client uses WebSockets instead, but both have the same path modulo protocol)
-        self.putChild(AUDIO_STREAM_PATH_ELEMENT, AudioStreamResource(session))
+        self.putChild(AUDIO_STREAM_PATH_ELEMENT.encode(), AudioStreamResource(session))
```
Same concerns as previous component.

quentinmit replied:
Done.

kpreid commented on shinysdr/i/network/webapp.py:
```diff
@@ -50,13 +51,43 @@
 from shinysdr.values import SubscriptionContext
 
 
+# txWS is unfortunately dead upstream, and has buggy Python 3 support.
```
:(

I have a dead branch that tries to move to Autobahn, but I was having trouble getting it to even show me any error stacks from my callbacks not working.

kpreid commented on shinysdr/i/network/webapp.py:
```diff
@@ -188,10 +219,10 @@ def _put_plugin_resources(wcommon, client_resource):
     for resource_def in getPlugins(_IClientResourceDef, shinysdr.plugins):
         # Add the plugin's resource to static serving
-        plugin_resources.putChild(resource_def.key, resource_def.resource)
+        plugin_resources.putChild(resource_def.key.encode(), resource_def.resource)
```
This one definitely needs to be 'utf-8' as the plugin could put arbitrary characters in its names.

quentinmit replied:
Does that mean it also needs to be URL-encoded? Otherwise done.

kpreid commented on shinysdr/i/test_db.py:
```diff
@@ -290,7 +290,7 @@ def test_index_common(self):
     def test_index_response(self):
         response, data = yield http_get(reactor, self.__url('/'))
-        self.assertEqual(response.headers.getRawHeaders('Content-Type'), ['application/json'])
+        self.assertEqual(response.headers.getRawHeaders('Content-Type'), ['application/json; charset=utf-8'])
```
No charset parameter, as previously described

kpreid commented on shinysdr/plugins/rtl_433.py:
```diff
@@ -97,7 +97,7 @@ def __init__(self, mode='433', input_rate=0, context=None):
                 b'env', b'rtl_433',
                 b'-F', b'json',  # output format
-                b'-r', str(demod_rate) + b'sps:iq:cf32:-',  # specify input format and to use stdin
+                b'-r', b'%dsps:iq:cf32:-' % (demod_rate),  # specify input format and to use stdin
```
Either add a trailing comma or remove the parens; right now the parens are doing nothing.

quentinmit replied:
Good catch; the trailing comma was my intention.

kpreid reviewed (changes requested):
Code review only; I haven't run these myself yet.

quentinmit committed:
Address review comments

quentinmit committed:
Test removing charset from application/json headers

quentinmit commented:
It's been three weeks, have you had a chance to get your GR development environment working yet?

kpreid commented:
Sorry, I've been very busy dealing with various further life complications, and also a bit of obsession with another project. This is still on my todo list.

alphafox02 commented:
I’ve been checking back here from time to time. I’d like to add shinysdr back to DragonOS (project of mine). It’s already got python3/gnuradio 3.8 included. Just curious if there’s any movement planned on this or if I should just try to piece it together based on this PR and test it that way? Thanks

alphafox02 commented:
Tested it real quick and found that it works rather well. Thanks for doing this PR. I went ahead and also installed gr-air-modes and gr-dad. I had pretty much everything else that was needed, except gr-radioteletype which looks to be for an older gnuradio. 

One thing I noticed, I can see mode S info in the terminal but am I supposed to be seeing the messages decoded somewhere in the shinysdr GUI? I have the map up and do not see anything there.

alphafox02 commented:
I think I figured out that for whatever reason I’m not getting positions on Adsb, but I do see mode S messages. 

Another note, I thought I’d get lucky and this would work,

https://github.com/mcapdeville/gr-radioteletype/commit/69b6b182e67a85ddce581037a10f176e3516d833

But after installing the 3.8 version shiny SDR would not start till I uninstalled gr-radioteletype. Thought maybe I’d have all the features working, so close.

NeuroForLunch commented:
Love this project. I hope this branch gets completed and this project upgraded to python 3!

revivec commented:
我现在正在开始的迁移工作，从gnuradio3.7+python2.7迁移到gnuradio3.9+python3.8  正在迁移的过程中，目前通过python setup.py build  和  sudo python setup.py install 返回的错误提示修改了两处位置。


1.  在setup.py 中  urllib改成了urllib.request
2. 关于rational_resampler
 #from gnuradio.filter import rational_resampler


try:
    from gnuradio.filter import rational_resampler
except ImportError:
    dirname, filename = os.path.split(os.path.abspath(__file__))
    __path__.append(os.path.join(dirname, "bindings"))
    from gnuradio.filter.filter_python import rational_resampler

并且升级了   Twisted six txWS PyEphem pyserial等相关依赖，但是目前还是存在问题。
是否有成功的给与我一些建议。谢谢。
`

s-martin commented:
Hi, are there any plans to merge this PR?

alphafox02 commented:
Be even better to see if support gr 3.9+

martindale commented:
Arriving here after an excursion installing the now-deprecated Python 2 on an older Debian-based OS, now looking into doing this upgrade myself to more seamlessly integrate with modern systems!
